src/main/java/org/spongepowered/api/item/recipe/RecipeRegistry.java

```diff
@@ -24,32 +24,37 @@
  */
 package org.spongepowered.api.item.recipe;
 
-import java.util.Set;
+import com.google.common.collect.ImmutableCollection;
+import org.spongepowered.api.item.recipe.crafting.CraftingRegistry;
+import org.spongepowered.api.item.recipe.smelting.SmeltingRegistry;
+
+import java.util.function.Predicate;
 
 /**
- * A RecipeRegistry holds all registered recipes for a given game.
+ * Used for {@link SmeltingRegistry} and {@link CraftingRegistry}
  */
-public interface RecipeRegistry {
+public interface RecipeRegistry<T extends Recipe> {
 
     /**
      * Registers the given Recipe to make it available to craft.
      *
      * @param recipe The Recipe to register
      */
-    void register(Recipe recipe);
+    void register(T recipe);
 
     /**
-     * Removes the given Recipe from registration in this registry.
+     * Removes Recipes from this Registry using the given Predicate.
      *
-     * @param recipe The Recipe to unregister
+     * @param predicate The check to remove the Recipe or not
+     * @return If it removed any recipes or not
      */
-    void remove(Recipe recipe);
+    boolean remove(Predicate<? super T> predicate);
 
     /**
      * Retrieves all recipes registered in this registry.
      *
      * @return All registered recipes
      */
-    Set<Recipe> getRecipes();
+    ImmutableCollection<T> getRecipes();
```

Comment:
Don't use `ImmutableCollection`, just `Collection` is fine (you can even say it's unmodifiable in the javadoc).